kernel/src/hil/display.rs

```diff
@@ -0,0 +1,334 @@
+            0 => Some(ScreenPixelFormat::Mono),
+            1 => Some(ScreenPixelFormat::RGB_233),
+            2 => Some(ScreenPixelFormat::RGB_565),
+            3 => Some(ScreenPixelFormat::RGB_888),
+            4 => Some(ScreenPixelFormat::ARGB_8888),
+            5 => Some(ScreenPixelFormat::RGB_4),
+            Self::CHOICES_MASK => Some(ScreenPixelFormat::Other((val >> 24) as u8)),
+            _ => None,
+        }
+    }
+}
+
+/// Describes a rectangular area in the frame buffer. Sized in pixels.
+pub struct Area {
+    /// First column
+    pub x: usize,
+    /// First row
+    pub y: usize,
+    /// Column count
+    pub width: usize,
+    /// Row count
+    pub height: usize,
+}
+
+/// A frame buffer.
+///
+/// The frame buffer may be stored on the device itself and inaccessible.
+pub trait FrameBuffer {
+    /// Returns a tuple (width, height) with the current resolution (in pixels)
+    /// This function is synchronous as the driver should know this value without
+    /// requesting it from the screen.
+    ///
+    /// note that width and height may change due to rotation
+    fn get_resolution(&self) -> (usize, usize);
+
+    /// Returns the current pixel format
+    /// This function is synchronous as the driver should know this value without
+    /// requesting it from the screen.
+    fn get_pixel_format(&self) -> ScreenPixelFormat;
+
+    /// Sets the video memory area.
+    ///
+    /// This function has to be called before the first call to the write function.
+    ///
+    /// Area's validity depends on the current pixel format:
+    /// - the first pixel of the first column must start on a byte boundary, and
+    /// - the last pixel of the last column must start on a byte boundary.
+    /// This is to reduce ambiguity regarding partial bytes.
+    ///
+    /// This will generate a `command_complete()` callback when finished.
+    ///
+    /// Return values:
+    /// - `Ok(())`: The write area is valid.
+    /// - `INVAL`: The parameters of the write area are not valid.
+    /// - `BUSY`: Unable to set the write area on the device.
+    fn set_write_area(&self, area: &Area) -> Result<(), ErrorCode>;
+
+    /// Sends a write command to write data in the selected video memory area.
+    /// When finished, the driver will call the `write_complete()` callback.
+    ///
+    /// Return values:
+    /// - `Ok(())`: Write is valid and will be sent to the screen.
+    /// - `RESERVE`: No write area reserved.
+    /// - `INVAL`: Write is invalid or length is wrong.
+    /// - `BUSY`: Another write is in progress.
+    fn write(&self, buffer: &'static mut [u8], len: usize) -> Result<(), ErrorCode>;
+
+    /// Applies all unapplied write commands.
+    ///
+    /// When finished, the driver shall call `ScreenClient::write_complete`.
+    /// Returns `BUSY` if another operation is in progress,
+    /// or `ENOSUPPORT` if the driver automatically applies all writes.
+    fn flush(&self) -> Result<(), ErrorCode> {
```

Comment:
I now remember the use case of `write_continue` and how I used it. 

The small screens that I used would reset the write window (write from its beginning) when a new SPI write command was issues. As screen data may be large (I have a 240 x 240 16 bit screen), the `ProcessBuffer` sent by an application may not fit into the driver's buffer. This is why I needed to distinguish between an actual write (that would reset the screen's write position) and continued write.

This mechanism could be implemented by using `write` and `flush`, but a driver might forget to use `flush` at the end of writes, which would prevent the screen from reseting the write position. I am leaning towards keeping al the three functions, but others might want to chip in. It is true that having only `write` and `flush` makes the API simpler, but does complicate the implementation and adds a little room for error.

Reply:
I think `flush` is not really an integral part of this API, so I'm willing to leave it out for later.

I don't think that `continue` is needed for your use case: if the device expects an *entire* frame at once, then the driver can size its buffer accordingly. A write request bigger than that will be more than one frame, and is an indication that something is wrong.

However, `continue` might be useful for situations where the device expects an entire buffer inside a single write transaction, but there's not enough RAM to hold it all.

Given that there's no way to submit such buffers today (SPI doesn't use LeasedBuffer), I'd rather drop `continue` and worry about this use case later.

Reply:
> I don't think that `continue` is needed for your use case: if the device expects an _entire_ frame at once, then the driver can size its buffer accordingly. A write request bigger than that will be more than one frame, and is an indication that something is wrong.
> 

I think I have to disagree on this. The problem is the discrepancy between the application buffer size and the syscall driver's buffer size. If the application buffer size is larger that the one of the syscall driver (as the application has no way of knowing the syscall driver's buffer size), the syscall driver has to issue *several writes* to the display driver that need to appear *as one* for the actual device.

Let's say we have a buffer of 100 bytes from the application and a buffer of 30 bytes available in the syscall driver. The actual device needs to receive all the 100 bytes in one single command. This is what happens:

`write (30 bytes)` -> sends WRITE_CMD BUFFER to the communication channel
`continue_write(30bytes)` sends BUFFER to the communication channel (not WRITE followed by BUFFER)
`continue_write(30bytes)` sends BUFFER to the communication channel (not WRITE followed by BUFFER)
`continue_write(10bytes)` sends BUFFER to the communication channel (not WRITE followed by BUFFER)

Where it not for `continue_write`, the syscall driver would send WRITE BUFFER every time, which would result in a reset of the write position. The only way to avoid this would be for the sycall driver to recalculate the write frame position, which seems a lot of overhead.

> However, `continue` might be useful for situations where the device expects an entire buffer inside a single write transaction, but there's not enough RAM to hold it all.
> 

This is mostly why the syscall driver might have a smaller buffer then the application.

> Given that there's no way to submit such buffers today (SPI doesn't use LeasedBuffer), I'd rather drop `continue` and worry about this use case later.

I'm not sure what you mean by this.

Reply:
No, we're in complete agreement :) I'm saying that:

> a buffer of 30 bytes available in the syscall driver.

is a bug. The syscall driver can either:

- split the user's buffer into squares, each of which is <30 bytes long, and submit those or
- allocate a buffer big enough to keep the entire frame of the underlying device driver, and reject writes bigger than that.

That thing with LeasedBuffer is actually probably more of a consequence of another problem. I'll address this later, but for now, I think that `continue` forces additional complexity that shouldn't be present while figuring out the general shape of the HIL.

Reply:
>     * split the user's buffer into squares, each of which is <30 bytes long, and submit those or
>
This is exactly what `continue`, it splits the buffer

>     * allocate a buffer big enough to keep the entire frame of the underlying device driver, and reject writes bigger than that.
> 
This is difficult to do, as applications have no idea (so far) about what that buffer is. My opinion is that these details should be completely abstracted from the application.

Reply:
Yes, but `continue` splits the buffer in a special, non-square way, whereas the `frame` API for submitting squares could be used instead, without adding more methods.

As for allocation, it's the system call driver that needs to allocate the big enough buffer, not the userspace. This would be hardcoded in the board file, and transparent to the userspace.

Reply:
> Yes, but `continue` splits the buffer in a special, non-square way, whereas the `frame` API for submitting squares could be used instead, without adding more methods.
> 
Are you suggesting to have a function that reports to the syscall driver the size of the *tile* buffer? This needs to be somehow propagated to the user space then.

> As for allocation, it's the system call driver that needs to allocate the big enough buffer, not the userspace. This would be hardcoded in the board file, and transparent to the userspace.

I assume the syscall driver would have to allocate a buffer with a size multiple of a *tile*, not necessarily the whole framebuffer.

Reply:
The `continue` problem is mostly independent of the tile problem. I just suggest using the same function just smaller area for when a buffer needs to be split up, without proposing how exactly.

> I assume the syscall driver would have to allocate a buffer with a size multiple of a tile, not necessarily the whole framebuffer.

I think that's mostly right. In my case, the buffer would be sized to 1 line, and that's my minimal tile too.

Reply:
This might add complexity to the display driver as it would have split the actual write frame into smaller parts and issue several SetFrame commands to the underlying driver. After writing, it should than reset the frame to the original frame that was requested by the application.

Reply:
Right, there's going to be complexity in this either way. But when it's unavoidable, I'd rather have it complex in a uniform way :)